rs/utils/thread/src/deallocator_thread.rs

```diff
@@ -0,0 +1,101 @@
+use crate::JoinOnDrop;
+use crossbeam_channel::{bounded, unbounded, Sender};
+
+/// An object to be deallocated in the background.
+type Deallocation = Box<dyn std::any::Any + Send + 'static>;
+
+/// A helper struct that notifies when it's deallocated.
+struct NotifyWhenDeallocated {
+    channel: Sender<()>,
+}
+
+impl Drop for NotifyWhenDeallocated {
+    fn drop(&mut self) {
+        self.channel
+            .send(())
+            .expect("Failed to notify deallocation");
+    }
+}
+
+// We will not use the deallocation thread when the number of pending
+// deallocation objects goes above the threshold.
+const DEALLOCATION_BACKLOG_THRESHOLD: usize = 500;
+
+/// A thread that deallocates complex objects in the background. It spreads the
+/// cost of deallocation over a longer period of time, to avoid long pauses.
+pub struct DeallocatorThread {
```

Comment:
I really like that this is encapsulated now! One question that crossed my mind: do we also need some tests?